configuration/config2/SchemaValue.hpp

```diff
@@ -0,0 +1,105 @@
+        void addModifer(const std::string &mod) { m_modifiers.push_back(mod); }
+        void setModifiers(const std::vector<std::string> &list) { m_modifiers = list;  }
+        const std::vector<std::string> &getModifiers() const { return m_modifiers; }
+        bool hasModifiers() const { return m_modifiers.size() != 0; }
+        void setUniqueValue(bool isUnique) { bitMask.m_isUnique = isUnique; }
+        bool isUniqueValue() const { return bitMask.m_isUnique;  }
+        void setUniqueValueSetRef(const std::shared_ptr<SchemaValue> &pValue) { m_pUniqueValueSetRefs.push_back(pValue);  }
+        bool isFromUniqueValueSet() const { return !m_pUniqueValueSetRefs.empty(); }
+        std::vector<std::weak_ptr<SchemaValue>> getUniqueValueSetRefs() const { return m_pUniqueValueSetRefs;  }
+        bool isDefined() const { return bitMask.m_isDefined;  }
+        void resetEnvironment();
+        void setMirrorFromPath(const std::string &path) { m_mirrorFromPath = path;  }
+        const std::string &getMirrorFromPath() const { return m_mirrorFromPath;  }
+        bool isMirroredValue() const { return m_mirrorFromPath.length() != 0; }
+        void addMirroredSchemaValue(const std::shared_ptr<SchemaValue> &pVal) { m_mirrorToSchemaValues.push_back(pVal); }
+        void mirrorValueToEnvironment(const std::string &oldValue, const std::string &newValue);
+        void addEnvironmentValue(const std::shared_ptr<EnvironmentValue> &pEnvValue) { m_envValues.push_back(pEnvValue); }
+        std::vector<std::string> getAllEnvironmentValues() const;
+        void setMirroredEnvironmentValues(const std::string &oldValue, const std::string &newValue);
+        void validate(Status &status, const std::string &id, const EnvironmentValue *pEnvValue = nullptr) const;
+        std::vector<AllowedValue> getAllowedValues(const EnvironmentValue *pEnvValue = nullptr) const;
+        void setAutoGenerateType(const std::string &type) { m_autoGenerateType = type; }
+        const std::string &getAutoGenerateType() const { return m_autoGenerateType; }
+        void setAutoGenerateValue(const std::string &value) { m_autoGenerateValue = value; }
+        const std::string &getAutoGenerateValue() const { return m_autoGenerateValue; }
+        std::vector<std::string> getAllKeyRefValues() const;
+
+
+    protected:
+
+        std::shared_ptr<SchemaType> m_pType;
+        std::vector<std::weak_ptr<EnvironmentValue>> m_envValues;
+        std::vector<std::shared_ptr<SchemaValue>> m_mirrorToSchemaValues;
+        std::string m_name;
+        std::string m_displayName;
+        std::string m_mirrorFromPath;
+        std::string m_autoGenerateValue;
+        std::string m_autoGenerateType;
+
+        struct {
+            unsigned m_required  : 1;
+            unsigned m_readOnly  : 1;
+            unsigned m_hidden    : 1;
+            unsigned m_deprecated: 1;
+            unsigned m_isUnique  : 1;
+            unsigned m_isDefined : 1;
+        } bitMask;
```

Comment:
I didn't notice this struct being used in any special way, is there any reason these couldn't be just a series  SchemaValue member vars of type bool? and if we wanted to create an actual bit mask, why not use a single unsigned to represent the bitmask (where each of the bits represent a different meaning)?

Reply:
This is a standard way of defining a bitmask where the members of the struct, bitMask, are single bits (or multiple bits if needed). The reason is to reduce memory usage. Multiple bools each take more memory that compacting all of the bool values into a single bitmask